color_refinement_helper.py

```diff
@@ -59,27 +59,27 @@ def has_same_color_neighbours(u: Vertex, v: Vertex, coloring: Coloring) -> bool:
     :param coloring: current coloring
     :return: `True` if the vertices have the same colored neighbourhood, `False` otherwise
     """
-    ncolors_u = [coloring.color(w) for w in u.neighbours]
-    ncolors_v = [coloring.color(w) for w in v.neighbours]
+    ncolors_u = (coloring.color(w) for w in u.neighbours)
+    ncolors_v = (coloring.color(w) for w in v.neighbours)
     return compare(ncolors_u, ncolors_v)
 
 
-def choose_color(coloring: Coloring) -> List[Vertex]:
+def choose_color(coloring: Coloring) -> DoubleLinkedList:
     """
     Returns a partition cell (aka color class) with at least four vertices
 
     Returns the first color class with at least four vertices that is found.
     :param coloring: current coloring
     :return: a color class with at least four vertices, `None` if no color class could be found
     """
-    for key in coloring.colors:
-        vertices = list(coloring.get(key))
+    for color in coloring.colors:
+        vertices = coloring.get(color)
         if len(vertices) >= 4 and len(vertices) % 2 == 0:
             return vertices
-    return []
+    return DoubleLinkedList()
```

Comment:
Door een echte `DoubleLinkedList` terug te geven behouden we hiermee nu de voordelen van die datastructuur.